File: JavaScript/src/_00070_ClimbingStairs.js
/**
 * Given `n` steps to reach the top of a staircase, find the number of distinct
 * ways you can climb to the top if each time you can either climb 1 or 2 steps.
 * @param {number} n number of steps
 * @returns {number} number of distinct ways to climb to the top of the stairs
 */
var climbStairs = function (n) {
    // If standing on any step `n`, one of two possibilities must have occured
    // in the last climb to get to this step. We were either on step `n - 1` and 
    // took one step, or we were on step `n - 2` and took two steps. The number
    // of ways to reach `n` is always the number of ways to reach `n - 1` and
    // the number of ways to reach `n - 2`.

    // Base case:
    // If there are 0 or 1 steps, there is exactly 1 way to reach the top.
    // (Either you're already there, or you take a single step.)
    if (n <= 1) return 1;

    // prev2 represents ways(i - 2)
    // Initially, ways(0) = 1
    let prev2 = 1; 

    // prev1 represents ways(i - 1)
    // Initially, ways(1) = 1
    let prev1 = 1;

    // Build the solution from step 2 up to step n
    for (let i = 2; i <= n; i++) {
        // The current number of ways is the sum of:
        // ways to reach previous step (i - 1)
        // plus ways to reach two steps before (i - 2)
        let current = prev1 + prev2;

        // Shift values forward for next iteration:
        // The old prev1 becomes prev2.
        prev2 = prev1;
        // The newly computed current becomes prev1.
        prev1 = current;
    }

    // After the loop, prev1 holds ways(n).
    return prev1;
}

/**
 * Given `n` steps to reach the top of a staircase, find the number of distinct
 * ways you can climb to the top if each time you can either climb 1 or 2 steps.
 * @param {number} n number of steps
 * @returns {number} number of distinct ways to climb to the top of the stairs
 */
var climbStairs1 = function (n) {
    // If standing on any step `n`, one of two possibilities must have occured
    // in the last climb to get to this step. We were either on step `n - 1` and 
    // took one step, or we were on step `n - 2` and took two steps. The number
    // of ways to reach `n` is always the number of ways to reach `n - 1` and
    // the number of ways to reach `n - 2`.
    if (n === 0 || n === 1) return 1;
    return climbStairs(n - 1) + climbStairs(n - 2);
};

module.exports = { climbStairs, climbStairs1 };

File: Python/_00070_ClimbingStairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        """Given `n` steps to reach the top of a staircase, find the number of 
        distinct ways you can climb to the top if each time you can either climb 
        1 or 2 steps.

        Args:
            n (int): number steps

        Returns:
            int: number of distinct ways to climb to the top of the stairs
        """

        # If there are one or no stairs, there is only one way to climb.
        if n <= 1:
            return 1
        
        prev2 = 1   # two climbs ago
        prev1 = 1   # one climb ago

        # For each step from the second step up...
        for i in range(2, n + 1):
            # the number of ways to climb to this step is the number of ways of
            # the previous two steps.
            current = prev1 + prev2
            # Remember the previous step into the next step.
            prev2 = prev1
            # Rememver this step into the next step.
            prev1 = current

        # When the loop ends, prev1 will have the ways for the last step.
        return prev1

import unittest

class TestSolution(unittest.TestCase):
    
    def setUp(self) -> None:
        self.sol = Solution()

    def test_example1(self):
        self.assertEqual(self.sol.climbStairs(2), 2)

    def test_example2(self):
        self.assertEqual(self.sol.climbStairs(3), 3)

    def test_greatest_input(self):
        self.assertEqual(self.sol.climbStairs(45), 1836311903)

if __name__ == '__main__':
    unittest.main()